### PR TITLE
Frontend: add sorting

### PR DESCRIPTION
**Pull request by danielbeardsley**
Add sorting to the frontend with an uncomplicated sort. I decided to abandon the score based sorting cause I feel it was overly complicated. I've gone for a simpler sort based on booleans:

* My pulls above pulls that aren't mind
* Pulls I have to CR/QA above those I don't
* Pulls I haven't touched above those I have already CRed

See commits for more info

### CODE REVIEW

File: frontend/src/pulldasher/pulls-context.tsx [hunk 1]
```diff
@@ -26,8 +28,25 @@ export function useSetFilter(): FilterFunctionSetter {
...
+   const [filteredPulls, setFilter] = useFilteredPullsState(unfilteredPulls);
+   const sortedPulls = filteredPulls.sort(defaultSort);
+   return (<PullsContext.Provider value={{pulls: sortedPulls, setFilter}}>
       {children}
    </PullsContext.Provider>);
 }
+
+function defaultSort(a: Pull, b: Pull): number {
+   return (
+    // My pulls above pulls that aren't mind
```
Comment (deltuh-vee) [inline frontend/src/pulldasher/pulls-context.tsx:41]:
```suggestion
    // My pulls above pulls that aren't mine
```

**Comment by danielbeardsley** [comment]
QA :+1: 

My pulls, then pulls I need to re-CR:
image

All the way at the bottom are pulls I've already CRed:
image

File: frontend/src/pulldasher/pulls-context.tsx [hunk 2]
```diff
@@ -26,8 +28,25 @@ export function useSetFilter(): FilterFunctionSetter {
...
 }
+
+function defaultSort(a: Pull, b: Pull): number {
+   return (
+    // My pulls above pulls that aren't mine
+    compareBool(a.isMine(), b.isMine()) ||
+    // Pulls I have to CR/QA above those I don't
+    compareBool(a.hasOutdatedSig(getUser()), b.hasOutdatedSig(getUser())) ||
+    // Pulls I haven't touched vs those I have already CRed
+    compareBool(!a.hasCurrentSig(getUser()), !b.hasCurrentSig(getUser()))
```
Comment (andyg0808) [inline frontend/src/pulldasher/pulls-context.tsx:46]: I wasn't confident that this worked, but I wrote up a spec in [Alloy](http://alloytools.org/) (attached below), and it looks like it should always satisfy the triangle inequality:

image

image


[pulldasher_sort_spec.zip](https://github.com/iFixit/pulldasher/files/8275773/pulldasher_sort_spec.zip)

**Review by andyg0808 — APPROVED** [review]

**Comment by andyg0808** [comment]
CR :goat: through 17c2e6d

**Comment by cdcline** [comment]
CR 👍

File: frontend/src/pulldasher/pulls-context.tsx [hunk 3]
```diff
@@ -26,8 +28,25 @@ export function useSetFilter(): FilterFunctionSetter {
...
+   );
+}
+
+function compareBool(a: boolean, b: boolean): number {
+   return a == b ? 0 : (a ? -1 : 1);
```
[the remark this answers is not in the document]
Comment (danielbeardsley) [reply frontend/src/pulldasher/pulls-context.tsx:51]: Sure seems like it ought to be faster, though that likely doesn't matter.

The double ternary isn't that easy to validate either, I'll go with yours.

File: frontend/src/pulldasher/pulls-context.tsx (hunk 2 above)
Comment (danielbeardsley) [reply frontend/src/pulldasher/pulls-context.tsx:46]: My intuition says "I can't imagine how it couldn't work".  The two equalities will still be 0 and the two inequalities will just reverse their values. Same affect as swapping the A for the B.

File: frontend/src/pulldasher/pulls-context.tsx [hunk 4]
```diff
@@ -26,8 +28,25 @@ export function useSetFilter(): FilterFunctionSetter {
 }
 
 export const PullsProvider = function({children}: {children: React.ReactNode}) {
-   const [pulls, setFilter] = useFilteredPullsState();
-   return (<PullsContext.Provider value={{pulls, setFilter}}>
+   const unfilteredPulls = usePullsState();
+   const [filteredPulls, setFilter] = useFilteredPullsState(unfilteredPulls);
+   const sortedPulls = filteredPulls.sort(defaultSort);
```
[the remark this answers is not in the document]
Comment (danielbeardsley) [reply frontend/src/pulldasher/pulls-context.tsx:33]: I see. And yes, I agree that the current code is safe and it's not worth adding in an array clone.

**Comment by danielbeardsley** [comment]
QA :+1: 

Sorting still operates just as before.

**Comment by danielbeardsley** [comment]
QA :+1:

**Comment by cdcline** [comment]
CR 👍

**Pull request merged**